src/hevm/src/EVM/CSE.hs

```diff
@@ -0,0 +1,110 @@
+{-# Language DataKinds #-}
+
+{- |
+    Module: EVM.CSE
+    Description: Common subexpression elimination for Expr ast
+-}
+
+module EVM.CSE (BufEnv, StoreEnv, eliminateExpr, eliminateFlat) where
+
+import Prelude hiding (Word, LT, GT)
+
+import Data.Map (Map)
+import qualified Data.Map as Map
+import Control.Monad.State
+import Data.Text (Text)
+import qualified Data.Text as T
+
+import EVM.Types
+import EVM.Traversals
+
+-- maps expressions to variable names
+data BuilderState = BuilderState
+  { bufs :: (Int, Map (Expr Buf) Int)
+  , stores :: (Int, Map (Expr Storage) Int)
+  }
+  deriving (Show)
+
+type BufEnv = Map (GVar Buf) (Expr Buf)
+type StoreEnv = Map (GVar Storage) (Expr Storage)
+
+initState :: BuilderState
+initState = BuilderState
+  { bufs = (0, Map.empty)
+  , stores = (0, Map.empty)
+  }
+
+
+go :: Expr a -> State BuilderState (Expr a)
+go = \case
+  -- buffers
+  e@(WriteWord i v b) -> do
```

Comment:
According to GHC, this can be replaced with `WriteWord {}`. Similarly below for `WriteByte,` `CopySlice`, and `SStore`. The `v` here shadows the `v` below in `Just v`, which confused me at first.

Reply:
Good point!

Reply:
Done!